grype/db/v6/distribution/client.go

```diff
@@ -0,0 +1,234 @@
+	defer downloadProgress.SetCompleted()
+
+	// note: as much as I'd like to use the afero FS abstraction here, the go-getter library does not support it
+	tempDir, err := os.MkdirTemp(dest, "grype-db-download")
+	if err != nil {
+		return "", fmt.Errorf("unable to create db temp dir: %w", err)
+	}
+
+	// download the db to the temp dir
+	u, err := url.Parse(c.config.LatestURL)
+	if err != nil {
+		removeAllOrLog(afero.NewOsFs(), tempDir)
+		return "", fmt.Errorf("unable to parse db URL %q: %w", c.config.LatestURL, err)
+	}
+
+	u.Path = path.Join(path.Dir(u.Path), path.Clean(archive.Path))
+
+	// from go-getter, adding a checksum as a query string will validate the payload after download
+	// note: the checksum query parameter is not sent to the server
+	query := u.Query()
+	if archive.Checksum != "" {
+		query.Add("checksum", archive.Checksum)
+	}
+	u.RawQuery = query.Encode()
+
+	// go-getter will automatically extract all files within the archive to the temp dir
+	err = c.updateDownloader.GetToDir(tempDir, u.String(), downloadProgress)
+	if err != nil {
+		removeAllOrLog(afero.NewOsFs(), tempDir)
+		return "", fmt.Errorf("unable to download db: %w", err)
+	}
+
+	return tempDir, nil
+}
+
+// latestFromURL loads a LatestDocument from a URL.
+func (c client) latestFromURL() (*LatestDocument, error) {
+	resp, err := c.latestHTTPClient.Get(c.config.LatestURL)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch latest.json: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unable to fetch latest.json: %s", resp.Status)
+	}
+
+	defer resp.Body.Close()
+
+	return NewLatestFromReader(resp.Body)
+}
+
+func withClientTimeout(timeout time.Duration) func(*http.Client) {
+	return func(c *http.Client) {
+		c.Timeout = timeout
+	}
+}
+
+func defaultHTTPClient(fs afero.Fs, caCertPath string, postProcessor ...func(*http.Client)) (*http.Client, error) {
+	httpClient := cleanhttp.DefaultClient()
+	httpClient.Timeout = 30 * time.Second
+	if caCertPath != "" {
+		rootCAs := x509.NewCertPool()
+
+		pemBytes, err := afero.ReadFile(fs, caCertPath)
+		if err != nil {
+			return nil, fmt.Errorf("unable to configure root CAs for curator: %w", err)
+		}
+		rootCAs.AppendCertsFromPEM(pemBytes)
+
+		httpClient.Transport.(*http.Transport).TLSClientConfig = &tls.Config{
+			MinVersion: tls.VersionTLS12,
+			RootCAs:    rootCAs,
+		}
+	}
+
+	for _, pp := range postProcessor {
+		pp(httpClient)
+	}
+
+	return httpClient, nil
+}
+
+func removeAllOrLog(fs afero.Fs, dir string) {
+	if err := fs.RemoveAll(dir); err != nil {
+		log.WithFields("error", err).Warnf("failed to remove path %q", dir)
+	}
+}
+
+func isSupersededBy(m *v6.Description, other v6.Description) bool {
+	if m == nil {
+		log.Debugf("cannot find existing metadata, using update...")
+		// any valid update beats no database, use it!
+		return true
+	}
+
+	otherModel, otherOk := other.SchemaVersion.Model()
```

Comment:
Can we rename or comment this? The `model` is the 6 of `6.0.0` but I thought it was going to be the whole `6.0.0` and misunderstood the rest of the function. I know we don't want `major` because that's semver, not schemaver, but `otherModelComponent` might be nice because it calls attention to the fact that this isn't the whole thing. (`Model` is a word that doesn't mean much due to overloading, like `instance` or `container` or something, so I didn't realize the method call on this line subsetted the SchemaVersion).

Maybe I just need to know schemaver better 🤷

Reply:
I renamed those helper functions to `*Part` (e.g. `ModelPart`)